src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs

```diff
@@ -405,5 +405,16 @@ public void Dispose()
             State = FunctionInvocationDispatcherState.Disposing;
             Dispose(true);
         }
+
+        public async Task RestartAsync()
+        {
+            // Dispose and restart all initialized channels
+            State = FunctionInvocationDispatcherState.WorkerProcessRestarting;
+            var channels = await GetInitializedWorkerChannelsAsync();
+            foreach (var channel in channels)
+            {
+                await DisposeAndRestartWorkerChannel(_workerRuntime, channel.Id);
```

Comment:
Maybe too big of a change for this PR, but we should probably also drain executions here before shutting down (because a timeout, unlike an unhandled exception on the process, could have other healthy executions running too). Maybe we file an issue as a separate improvement (application shutdown has graceful timeouts on executions, so we are giving some of that up).

Reply:
We already have a grace period in the timeout exception itself though.
![image](https://user-images.githubusercontent.com/26885664/76896522-363e9e80-684f-11ea-8f18-153203a9f649.png)

Reply:
I think new requests are still going to be routed to the worker though? So even though there's the 2 second grace period, existing requests are finishing and new requests are just starting to execute?

Reply:
I think that's the behavior today as well even with normal shutdown. Even if we give grace periods for execution to complete, as long as the web app itself is accepting requests, there are chances some additional executions might flow into the worker/channel shutting down..

Reply:
@yojagad  - wasn't the change you made to add [DrainInvocationsAsync](https://github.com/Azure/azure-functions-host/commit/4a9aa46e9442ea9d4b0b68fbe6d5c730b8c90baf) in order to complete outstanding executions, while not allowing new ones (because the jobhost is shut down at that point)? Or maybe we need to add `State = Stopped` or something to our StopAsync to make that happen.

but either way! since we're just recycling workers we should be able to do better i think..!